doc/source/ray-core/examples/batch_prediction.ipynb

```diff
@@ -0,0 +1,283 @@
+    "results = []\n",
+    "for file in input_files:\n",
+    "    if len(result_refs) > max_in_flight_tasks:\n",
+    "        num_ready = len(result_refs) - max_in_flight_tasks\n",
+    "        newly_completed, result_refs = ray.wait(result_refs, num_returns=num_ready)\n",
+    "        for completed_ref in newly_completed:\n",
+    "            results.append(ray.get(completed_ref))\n",
+    "    result_refs.append(make_prediction.remote(model_ref, file))\n",
+    "\n",
+    "newly_completed, result_refs = ray.wait(result_refs, num_returns=len(result_refs))\n",
+    "for completed_ref in newly_completed:\n",
+    "    results.append(ray.get(completed_ref))\n",
+    "for r in results:\n",
+    "    print(\"r size:\", r)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "```{tip}\n",
+    "To avoid repeatedly storing the same model into object store (this can cause Out-of-disk for driver node), use ray.put() to store the model once, and then pass the reference around\n",
+    "```\n",
+    "```{tip}\n",
+    "To avoid congest or overload the driver node, it’s preferable to have each task to write out the predictions (instead of returning results back to driver which will then do nothing but write out to storage system)\n",
+    "```\n",
+    "\n",
+    "## Actor-based batch prediction\n",
+    "In the above solution, each Ray task will have to fetch the model from the driver node before it can start performing the prediction. This is an overhead cost that can be significant if the model size is large. We can optimize it by using Ray actors, which will fetch the model just once and reuse it for all tasks assigned to the actor.\n",
+    "\n",
+    "First, we define a callable class that’s structured with an interface (i.e. constructor) to load/cache the model, and the other to take in a file and perform prediction."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import pandas as pd\n",
+    "import pyarrow.parquet as pq\n",
+    "import ray\n",
+    "\n",
+    "@ray.remote\n",
+    "class BatchPredictor:\n",
+    "    def __init__(self, model):\n",
+    "        self.model = model\n",
+    "        \n",
+    "    def predict(self, shard_path) -> pd.DataFrame:\n",
+    "        df = pq.read_table(shard_path).to_pandas()\n",
+    "        result =self.model(df)\n",
+    "        # Write the prediction result out.\n",
+    "        # Here we just return the size about the result.\n",
+    "        return len(result)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The constructor is called only once per actor worker. We use ActorPool to manage a set of actors that can receive prediction requests."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from ray.util.actor_pool import ActorPool\n",
```

Comment:
Hmm, ActorPool is deprecated in favor of using Datasets.map_batches

Reply:
Yes, this gets a bit tricky. Actually map_batches() is the recommended way to run batch prediction.

In retrospect, we probably should keep ActorPool.